test/run/test_stages.py

```diff
@@ -537,12 +537,13 @@ def test_dnf4_mark(self):
             assert os.path.isdir(tree)
 
             # we're going to verify that packages in the tree are marked according to
-            # Explicitly use 'dnf4' for now, because 'dnf5' contains a breaking change
+            # Explicitly use 'dnf-3' for now, because 'dnf5' contains a breaking change
             # in the repoquery --qf output, specifically it does not add a trailing newline.
             # We can use plan 'dnf' again once https://github.com/rpm-software-management/dnf5/issues/709 is fixed.
+            # We should be ideally using 'dnf4' here, but the '/usr/bin/dnf4 -> dnf-3' symlink does not exist on c9s.
             r = subprocess.run(
                 [
-                    "dnf4",
+                    "dnf-3",
```

Comment:
Thanks for digging into this, would tweaking the test to support both the (broken) and correct output help maybe? Something like:
```python
diff --git a/test/run/test_stages.py b/test/run/test_stages.py
index ca224f07..91eece3f 100644
--- a/test/run/test_stages.py
+++ b/test/run/test_stages.py
@@ -542,10 +542,10 @@ class TestStages(test.TestBase):
             # We can use plan 'dnf' again once https://github.com/rpm-software-management/dnf5/issues/709 is fixed.
             r = subprocess.run(
                 [
-                    "dnf4",
+                    "dnf",
                     "--installroot", tree,
                     "repoquery", "--installed",
-                    "--qf", "%{name},%{reason}"
+                    "--qf", "%{name},%{reason}\n"
                 ],
                 stdout=subprocess.PIPE,
                 stderr=subprocess.PIPE,
@@ -554,6 +554,8 @@ class TestStages(test.TestBase):
             )
 
             for line in r.stdout.splitlines():
+                if not line:
+                    continue
                 package, mark = line.strip().split(",")
 
                 if package == "dnf":
```
(and updating the comment)? But I guess its fine either way assuming the upstream bug gets fixed relatively soon (the advantage of having it handle both would be that even if we forget to update the code we would still test dnf4 and dnf5, with forcing dnf4 we would leave a small testing gap (of course only if we forget about the update :))

Reply:
Thanks for the suggestion. Personally, I decided that it doesn't make sense to handle both situations, since the upstream acknowledged the issue and plans to fix it.

WRT the testing gap - the intention is IMO not to test both dnf4 and dnf5 here, because:

- the `dnf` binary is used merely as a way to inspect the fs tree.
- the test case would always use a single version of the `dnf` binary from the host. So there would be always potentially one version that ends up not being tested.
- the `dnf` version that was used to mark the installed packages (from the buildroot pipeline) may be different than the one that is available on the host.

Nevertheless, I'm not opposed to your solution... It won't require "fixing" the test case when Fedora starts shipping only dnf5.

Reply:
Btw, my comment was in response to @mvo5's initial comment.  I didn't see @thozza's response until after I wrote it.